File: sdks/typescript/tests/schema-evolution.test.ts
/**
 * Unit tests for the SDK schema-evolution + observability surface (phase14):
 * - CollectionsClient: renameCollection, reindexCollection,
 *   snapshotCollectionNative, listCollectionSnapshotsNative,
 *   restoreCollectionSnapshotNative
 * - SearchClient: explainSearch
 * - AdminClient: listSlowQueries, setSlowQueryConfig
 *
 * Verifies wire shape (URL + body) and that server responses decode into
 * the expected model types. Pattern mirrors tests/tier-control.test.ts.
 */

import { describe, it, expect, beforeEach, vi } from 'vitest';
import { CollectionsClient } from '../src/client/collections';
import { SearchClient } from '../src/client/search';
import { AdminClient } from '../src/client/admin';
import type {
  ExplainResponse,
  NativeSnapshotInfo,
  ReindexJob,
  SlowQueryConfig,
  SlowQueryEntry,
} from '../src/models/schema-evolution';

// ---------------------------------------------------------------------------
// Shared mock transport helper
// ---------------------------------------------------------------------------

interface MockTransport {
  get: ReturnType<typeof vi.fn>;
  post: ReturnType<typeof vi.fn>;
  put: ReturnType<typeof vi.fn>;
  patch: ReturnType<typeof vi.fn>;
  delete: ReturnType<typeof vi.fn>;
  postFormData: ReturnType<typeof vi.fn>;
}

function createMock(): MockTransport {
  return {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
    patch: vi.fn(),
    delete: vi.fn(),
    postFormData: vi.fn(),
  };
}

// ---------------------------------------------------------------------------
// CollectionsClient — phase14
// ---------------------------------------------------------------------------

describe('CollectionsClient — schema-evolution (phase14)', () => {
  let mock: MockTransport;
  let client: CollectionsClient;

  beforeEach(() => {
    mock = createMock();
    client = new CollectionsClient({ transport: mock as never });
  });

  // ── renameCollection ──────────────────────────────────────────────────────

  describe('renameCollection', () => {
    it('POSTs to /collections/{name}/rename with new_name in body', async () => {
      mock.post.mockResolvedValue({ old_name: 'docs', new_name: 'docs_v2', alias_retained: 'docs', status: 'ok' });
      await client.renameCollection('docs', 'docs_v2');
      expect(mock.post).toHaveBeenCalledWith(
        '/collections/docs/rename',
        { new_name: 'docs_v2' },
      );
    });

    it('resolves void on success', async () => {
      mock.post.mockResolvedValue({ status: 'ok' });
      await expect(client.renameCollection('col', 'col2')).resolves.toBeUndefined();
    });
  });

  // ── reindexCollection ─────────────────────────────────────────────────────

  describe('reindexCollection', () => {
    const serverResponse: ReindexJob = {
      job_id: 'reindex-docs-1746000000',
      collection: 'docs',
      state: 'completed',
      params: { m: 32, ef_construction: 400, ef_search: 200 },
      progress: 1.0,
    };

    it('POSTs correct body to /collections/{name}/reindex', async () => {
      mock.post.mockResolvedValue(serverResponse);
      await client.reindexCollection('docs', { m: 32, ef_construction: 400, ef_search: 200 });
      expect(mock.post).toHaveBeenCalledWith(
        '/collections/docs/reindex',
        { m: 32, ef_construction: 400, ef_search: 200 },
      );
    });

    it('returns a ReindexJob with state "completed"', async () => {
      mock.post.mockResolvedValue(serverResponse);
      const job = await client.reindexCollection('docs', { m: 32, ef_construction: 400, ef_search: 200 });
      expect(job.state).toBe('completed');
      expect(job.job_id).toContain('reindex');
      expect(job.progress).toBe(1.0);
    });
  });

  // ── snapshotCollectionNative ──────────────────────────────────────────────

  describe('snapshotCollectionNative', () => {
    const snapshotResponse: NativeSnapshotInfo & { status: string } = {
      id: 'snap-abc-123',
      collection: 'docs',
      created_at: '2026-05-02T00:00:00Z',
      size_bytes: 4096,
      status: 'ok',
    };

    it('POSTs empty body to /collections/{name}/snapshot', async () => {
      mock.post.mockResolvedValue(snapshotResponse);
      await client.snapshotCollectionNative('docs');
      expect(mock.post).toHaveBeenCalledWith('/collections/docs/snapshot', {});
    });

    it('returns NativeSnapshotInfo with correct fields', async () => {
      mock.post.mockResolvedValue(snapshotResponse);
      const info = await client.snapshotCollectionNative('docs');
      expect(info.id).toBe('snap-abc-123');
      expect(info.collection).toBe('docs');
      expect(info.size_bytes).toBe(4096);
      expect(info.created_at).toBe('2026-05-02T00:00:00Z');
    });
  });

  // ── listCollectionSnapshotsNative ─────────────────────────────────────────

  describe('listCollectionSnapshotsNative', () => {
    const listResponse = {
      collection: 'docs',
      snapshots: [
        { id: 'snap-1', collection: 'docs', created_at: '2026-05-02T00:00:00Z', size_bytes: 2048 },
        { id: 'snap-2', collection: 'docs', created_at: '2026-05-02T01:00:00Z', size_bytes: 3000 },
      ],
      total: 2,
    };

    it('GETs /collections/{name}/snapshots', async () => {
      mock.get.mockResolvedValue(listResponse);
      await client.listCollectionSnapshotsNative('docs');
      expect(mock.get).toHaveBeenCalledWith('/collections/docs/snapshots');
    });

    it('returns the snapshots array', async () => {
      mock.get.mockResolvedValue(listResponse);
      const snaps = await client.listCollectionSnapshotsNative('docs');
      expect(snaps).toHaveLength(2);
      expect(snaps[0].id).toBe('snap-1');
      expect(snaps[1].size_bytes).toBe(3000);
    });

    it('returns empty array when snapshots missing', async () => {
      mock.get.mockResolvedValue({ collection: 'docs', total: 0 });
      const snaps = await client.listCollectionSnapshotsNative('docs');
      expect(snaps).toEqual([]);
    });
  });

  // ── restoreCollectionSnapshotNative ──────────────────────────────────────

  describe('restoreCollectionSnapshotNative', () => {
    it('POSTs empty body to /collections/{name}/snapshots/{id}/restore', async () => {
      mock.post.mockResolvedValue({ collection: 'docs', snapshot_id: 'snap-1', status: 'restored' });
      await client.restoreCollectionSnapshotNative('docs', 'snap-1');
      expect(mock.post).toHaveBeenCalledWith(
        '/collections/docs/snapshots/snap-1/restore',
        {},
      );
    });

    it('resolves void on success', async () => {
      mock.post.mockResolvedValue({ status: 'restored' });
      await expect(client.restoreCollectionSnapshotNative('docs', 'snap-1')).resolves.toBeUndefined();
    });
  });
});

// ---------------------------------------------------------------------------
// SearchClient — phase14
// ---------------------------------------------------------------------------

describe('SearchClient — explainSearch (phase14)', () => {
  let mock: MockTransport;
  let client: SearchClient;

  beforeEach(() => {
    mock = createMock();
    client = new SearchClient({ transport: mock as never });
  });

  const explainResponse: ExplainResponse = {
    collection: 'docs',
    k: 10,
    results: [{ id: 'vec-1', score: 0.95, payload: null }],
    trace: {
      visited_nodes: 120,
      ef_search: 100,
      hnsw_search_ms: 1.23,
      payload_filter_evals: 0,
      quantization_score_ms: 0.45,
      total_ms: 2.10,
    },
  };

  it('POSTs vector and k to /collections/{name}/explain', async () => {
    mock.post.mockResolvedValue(explainResponse);
    await client.explainSearch('docs', [0.1, 0.2, 0.3], 10);
    expect(mock.post).toHaveBeenCalledWith(
      '/collections/docs/explain',
      { vector: [0.1, 0.2, 0.3], k: 10 },
    );
  });

  it('omits k from body when not provided', async () => {
    mock.post.mockResolvedValue(explainResponse);
    await client.explainSearch('docs', [0.1, 0.2]);
    const [, body] = mock.post.mock.calls[0] as [string, Record<string, unknown>];
    expect(body).not.toHaveProperty('k');
    expect(body['vector']).toEqual([0.1, 0.2]);
  });

  it('returns ExplainResponse with all trace fields', async () => {
    mock.post.mockResolvedValue(explainResponse);
    const resp = await client.explainSearch('docs', [0.1, 0.2, 0.3], 10);
    expect(resp.collection).toBe('docs');
    expect(resp.k).toBe(10);
    expect(resp.results).toHaveLength(1);
    expect(resp.trace.visited_nodes).toBe(120);
    expect(resp.trace.ef_search).toBe(100);
    expect(resp.trace.hnsw_search_ms).toBeCloseTo(1.23);
    expect(resp.trace.payload_filter_evals).toBe(0);
    expect(resp.trace.quantization_score_ms).toBeCloseTo(0.45);
    expect(resp.trace.total_ms).toBeCloseTo(2.10);
  });
});

// ---------------------------------------------------------------------------
// AdminClient — phase14
// ---------------------------------------------------------------------------

describe('AdminClient — slow queries (phase14)', () => {
  let mock: MockTransport;
  let client: AdminClient;

  beforeEach(() => {
    mock = createMock();
    client = new AdminClient({ transport: mock as never });
  });

  // ── listSlowQueries ───────────────────────────────────────────────────────

  describe('listSlowQueries', () => {
    const serverResponse = {
      entries: [
        { timestamp: '2026-05-02T00:01:00Z', collection: 'docs', k: 10, duration_ms: 312.5 },
        { timestamp: '2026-05-02T00:02:00Z', collection: 'logs', k: 20, duration_ms: 800.0 },
      ] satisfies SlowQueryEntry[],
      total: 2,
      config: { threshold_ms: 200, capacity: 1000 } satisfies SlowQueryConfig,
    };

    it('GETs /slow_queries', async () => {
      mock.get.mockResolvedValue(serverResponse);
      await client.listSlowQueries();
      expect(mock.get).toHaveBeenCalledWith('/slow_queries');
    });

    it('returns the entries array', async () => {
      mock.get.mockResolvedValue(serverResponse);
      const entries = await client.listSlowQueries();
      expect(entries).toHaveLength(2);
      expect(entries[0].collection).toBe('docs');
      expect(entries[0].duration_ms).toBe(312.5);
      expect(entries[1].k).toBe(20);
    });

    it('returns empty array when entries missing', async () => {
      mock.get.mockResolvedValue({ total: 0, config: { threshold_ms: 200, capacity: 1000 } });
      const entries = await client.listSlowQueries();
      expect(entries).toEqual([]);
    });
  });

  // ── setSlowQueryConfig ────────────────────────────────────────────────────

  describe('setSlowQueryConfig', () => {
    const configResponse = { threshold_ms: 150, capacity: 500, status: 'ok' };

    it('POSTs correct body to /slow_queries/config', async () => {
      mock.post.mockResolvedValue(configResponse);
      await client.setSlowQueryConfig({ threshold_ms: 150, capacity: 500 });
      expect(mock.post).toHaveBeenCalledWith(
        '/slow_queries/config',
        { threshold_ms: 150, capacity: 500 },
      );
    });

    it('returns SlowQueryConfig without the server-only status field', async () => {
      mock.post.mockResolvedValue(configResponse);
      const result = await client.setSlowQueryConfig({ threshold_ms: 150, capacity: 500 });
      expect(result.threshold_ms).toBe(150);
      expect(result.capacity).toBe(500);
      expect((result as Record<string, unknown>)['status']).toBeUndefined();
    });
  });
});

File: sdks/go/schema_evolution_test.go
package vectorizer

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSchemaRenameCollection(t *testing.T) {
	var capturedBody map[string]string

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/my-col/rename" {
			t.Errorf("should call /collections/my-col/rename but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body but decode failed: %v", err)
		}

		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	err := client.RenameCollection("my-col", "my-col-v2")
	if err != nil {
		t.Fatalf("RenameCollection failed: %v", err)
	}

	if capturedBody["new_name"] != "my-col-v2" {
		t.Errorf("POST body should contain new_name=my-col-v2 but got %v", capturedBody["new_name"])
	}
}

func TestSchemaReindexCollection(t *testing.T) {
	var capturedBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/test-col/reindex" {
			t.Errorf("should call /collections/test-col/reindex but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body but decode failed: %v", err)
		}

		resp := ReindexJob{
			JobID:      "job-001",
			Collection: "test-col",
			State:      "completed",
			Params: map[string]interface{}{
				"m":               float64(16),
				"ef_construction": float64(200),
				"ef_search":       float64(100),
			},
			Progress: 1.0,
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	params := &ReindexParams{M: 16, EfConstruction: 200, EfSearch: 100}
	job, err := client.ReindexCollection("test-col", params)
	if err != nil {
		t.Fatalf("ReindexCollection failed: %v", err)
	}
	if job == nil {
		t.Fatal("should return non-nil ReindexJob")
	}
	if job.JobID != "job-001" {
		t.Errorf("should decode job_id as job-001 but got %s", job.JobID)
	}
	if job.State != "completed" {
		t.Errorf("should decode state as completed but got %s", job.State)
	}
	if job.Collection != "test-col" {
		t.Errorf("should decode collection as test-col but got %s", job.Collection)
	}

	if capturedBody["m"] != float64(16) {
		t.Errorf("POST body should contain m=16 but got %v", capturedBody["m"])
	}
	if capturedBody["ef_construction"] != float64(200) {
		t.Errorf("POST body should contain ef_construction=200 but got %v", capturedBody["ef_construction"])
	}
	if capturedBody["ef_search"] != float64(100) {
		t.Errorf("POST body should contain ef_search=100 but got %v", capturedBody["ef_search"])
	}
}

func TestSchemaSnapshotCollectionNative(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/my-col/snapshot" {
			t.Errorf("should call /collections/my-col/snapshot but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		resp := NativeSnapshotInfo{
			ID:         "snap-abc",
			Collection: "my-col",
			CreatedAt:  "2026-05-03T00:00:00Z",
			SizeBytes:  4096,
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	info, err := client.SnapshotCollectionNative("my-col")
	if err != nil {
		t.Fatalf("SnapshotCollectionNative failed: %v", err)
	}
	if info == nil {
		t.Fatal("should return non-nil NativeSnapshotInfo")
	}
	if info.ID != "snap-abc" {
		t.Errorf("should decode id as snap-abc but got %s", info.ID)
	}
	if info.Collection != "my-col" {
		t.Errorf("should decode collection as my-col but got %s", info.Collection)
	}
	if info.SizeBytes != 4096 {
		t.Errorf("should decode size_bytes as 4096 but got %d", info.SizeBytes)
	}
}

func TestSchemaListCollectionSnapshotsNative(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/my-col/snapshots" {
			t.Errorf("should call /collections/my-col/snapshots but got %s", r.URL.Path)
		}
		if r.Method != "GET" {
			t.Errorf("should use GET method but got %s", r.Method)
		}

		type envelope struct {
			Snapshots []NativeSnapshotInfo `json:"snapshots"`
		}
		resp := envelope{
			Snapshots: []NativeSnapshotInfo{
				{ID: "snap-1", Collection: "my-col", CreatedAt: "2026-05-03T00:00:00Z", SizeBytes: 1024},
				{ID: "snap-2", Collection: "my-col", CreatedAt: "2026-05-03T00:01:00Z", SizeBytes: 2048},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	snapshots, err := client.ListCollectionSnapshotsNative("my-col")
	if err != nil {
		t.Fatalf("ListCollectionSnapshotsNative failed: %v", err)
	}
	if len(snapshots) != 2 {
		t.Fatalf("should unwrap snapshots envelope and return 2 items but got %d", len(snapshots))
	}
	if snapshots[0].ID != "snap-1" {
		t.Errorf("first snapshot should have ID snap-1 but got %s", snapshots[0].ID)
	}
	if snapshots[1].ID != "snap-2" {
		t.Errorf("second snapshot should have ID snap-2 but got %s", snapshots[1].ID)
	}
}

func TestSchemaRestoreCollectionSnapshotNative(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/my-col/snapshots/snap-abc/restore" {
			t.Errorf("should call /collections/my-col/snapshots/snap-abc/restore but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	err := client.RestoreCollectionSnapshotNative("my-col", "snap-abc")
	if err != nil {
		t.Fatalf("RestoreCollectionSnapshotNative failed: %v", err)
	}
}

func TestSchemaExplainSearch(t *testing.T) {
	var capturedBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/collections/my-col/explain" {
			t.Errorf("should call /collections/my-col/explain but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body but decode failed: %v", err)
		}

		resp := ExplainResponse{
			Collection: "my-col",
			K:          5,
			Results:    []map[string]interface{}{},
			Trace: ExplainTrace{
				VisitedNodes:        42,
				EfSearch:            100,
				HnswSearchMs:        1.5,
				PayloadFilterEvals:  0,
				QuantizationScoreMs: 0.3,
				TotalMs:             2.1,
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	vector := []float32{0.1, 0.2, 0.3}
	result, err := client.ExplainSearch("my-col", vector, 5)
	if err != nil {
		t.Fatalf("ExplainSearch failed: %v", err)
	}
	if result == nil {
		t.Fatal("should return non-nil ExplainResponse")
	}
	if result.Collection != "my-col" {
		t.Errorf("should decode collection as my-col but got %s", result.Collection)
	}
	if result.K != 5 {
		t.Errorf("should decode k as 5 but got %d", result.K)
	}
	if result.Trace.VisitedNodes != 42 {
		t.Errorf("should decode trace.visited_nodes as 42 but got %d", result.Trace.VisitedNodes)
	}
	if result.Trace.TotalMs != 2.1 {
		t.Errorf("should decode trace.total_ms as 2.1 but got %f", result.Trace.TotalMs)
	}

	if capturedBody["k"] != float64(5) {
		t.Errorf("POST body should contain k=5 but got %v", capturedBody["k"])
	}
	rawVec, ok := capturedBody["vector"].([]interface{})
	if !ok {
		t.Fatalf("POST body should contain a vector array but got %T", capturedBody["vector"])
	}
	if len(rawVec) != 3 {
		t.Errorf("POST body vector should have 3 elements but got %d", len(rawVec))
	}
}

func TestSchemaListSlowQueries(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/slow_queries" {
			t.Errorf("should call /slow_queries but got %s", r.URL.Path)
		}
		if r.Method != "GET" {
			t.Errorf("should use GET method but got %s", r.Method)
		}

		type envelope struct {
			Entries []SlowQueryEntry `json:"entries"`
		}
		resp := envelope{
			Entries: []SlowQueryEntry{
				{Timestamp: "2026-05-03T00:00:00Z", Collection: "col-a", K: 10, DurationMs: 150.5},
				{Timestamp: "2026-05-03T00:01:00Z", Collection: "col-b", K: 5, DurationMs: 200.0},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	entries, err := client.ListSlowQueries()
	if err != nil {
		t.Fatalf("ListSlowQueries failed: %v", err)
	}
	if len(entries) != 2 {
		t.Fatalf("should unwrap entries envelope and return 2 items but got %d", len(entries))
	}
	if entries[0].Collection != "col-a" {
		t.Errorf("first entry should have collection col-a but got %s", entries[0].Collection)
	}
	if entries[0].DurationMs != 150.5 {
		t.Errorf("first entry should have duration_ms 150.5 but got %f", entries[0].DurationMs)
	}
	if entries[1].Collection != "col-b" {
		t.Errorf("second entry should have collection col-b but got %s", entries[1].Collection)
	}
}

func TestSchemaSetSlowQueryConfig(t *testing.T) {
	var capturedBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/slow_queries/config" {
			t.Errorf("should call /slow_queries/config but got %s", r.URL.Path)
		}
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body but decode failed: %v", err)
		}

		resp := SlowQueryConfig{
			ThresholdMs: 500,
			Capacity:    200,
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	cfg := &SlowQueryConfig{ThresholdMs: 500, Capacity: 200}
	result, err := client.SetSlowQueryConfig(cfg)
	if err != nil {
		t.Fatalf("SetSlowQueryConfig failed: %v", err)
	}
	if result == nil {
		t.Fatal("should return non-nil SlowQueryConfig")
	}
	if result.ThresholdMs != 500 {
		t.Errorf("should decode threshold_ms as 500 but got %d", result.ThresholdMs)
	}
	if result.Capacity != 200 {
		t.Errorf("should decode capacity as 200 but got %d", result.Capacity)
	}

	if capturedBody["threshold_ms"] != float64(500) {
		t.Errorf("POST body should contain threshold_ms=500 but got %v", capturedBody["threshold_ms"])
	}
	if capturedBody["capacity"] != float64(200) {
		t.Errorf("POST body should contain capacity=200 but got %v", capturedBody["capacity"])
	}
}
